starport-core/src/main/scala/com/krux/starport/system/DispatcherActor.scala

```diff
@@ -0,0 +1,106 @@
+package com.krux.starport.system
+
+import java.time.LocalDateTime
+
+import akka.actor.typed.scaladsl._
+import akka.actor.typed.{ActorRef, Behavior}
+import com.codahale.metrics.Timer
+import slick.jdbc.PostgresProfile.api._
+
+import com.krux.hyperion.expression.{Duration => HDuration}
+import com.krux.starport.config.StarportSettings
+import com.krux.starport.db.record.{Pipeline, ScheduledPipeline}
+import com.krux.starport.db.table.ScheduledPipelines
+import com.krux.starport.db.table.{Pipelines, ScheduleFailureCounters}
+import com.krux.starport.db.WaitForIt
+import com.krux.starport.exception.StarportException
+import com.krux.starport.util.DateTimeFunctions._
+import com.krux.starport.util.{DataPipelineDeploy, ErrorHandler}
+
+object DispatcherActor {
+
+  trait Msg
+
+  case class ScheduleReq(
+    pipeline: Pipeline,
+    scheduledStart: LocalDateTime,
+    scheduledEnd: LocalDateTime,
+    actualStart: LocalDateTime,
+    localJar: String,
+    timerContext: Timer.Context,
+    replyTo: ActorRef[ScheduleRsp]
+  ) extends Msg
+
+  case class ScheduleRsp(
+    scheduledPipelines: Either[StarportException, Seq[ScheduledPipeline]],
+    dispatcherAct: ActorRef[Msg]
+  )
+
+  def apply(starportSetting: StarportSettings): Behavior[Msg] =
+    Behaviors.setup(new DispatcherActor(_, starportSetting))
+
+}
+
+class DispatcherActor(context: ActorContext[DispatcherActor.Msg], starportSetting: StarportSettings)
+    extends AbstractBehavior[DispatcherActor.Msg](context) with WaitForIt {
+
+  import DispatcherActor._
+
+  override def onMessage(msg: Msg): Behavior[Msg] = msg match {
+    case ScheduleReq(
+          pipeline,
+          scheduledStart,
+          scheduledEnd,
+          actualStart,
+          localJar,
+          timerContext,
+          replyTo
+        ) =>
+      context.log.info(s"${context.self} dispatching pipline ${pipeline.name} (${pipeline.id})")
+      val result = DataPipelineDeploy.deployAndActive(
```

Comment:
@realstraw this is a blocking call because we need to wait on the datapipeline api, correct?

I don't know too much about akka, but my understanding is that you need to dedicate separate resources at the system level to dispatchers that wrap blocking calls, e.g. https://doc.akka.io/docs/akka/current/typed/dispatchers.html#blocking-needs-careful-management

I might be missing something but it looks like you scheduler and your dispatcher here would be pulling from the same pool of threads, which could impact scheduling availability.

Reply:
parallel processing is handled https://github.com/krux/starport/pull/61/files#diff-52baee92bb0604698a75577e091afbadR42 each dispatcher is supposed to be blocking, this is the same behavior as before

Reply:
Regarding thread pools, I'm not too concerned with thread starvation for now. These are low volume messages, plus these actors are not designed to handle new messages while performing scheduling (see the link above for implementation details). I think adding more execution context would introduce more complexity without much benefit.

Reply:
Ah -- so is never the case that there are more than one `SchedulerActor`s spawned at the same time?

Reply:
in this case we won't have more than one `SchedulerActor` at the same time, but even though there are multiple `SchedulerActor`s, one `DispatcherActor` still would not receive messages while it's working.